### PR TITLE
Pattern matching performance

### PR DESCRIPTION
**Pull request by fcevado**

### CODE REVIEW

**Comment by fcevado** [comment]
Changes only on the `Phone.Countries` module results:
```
## PhoneBench
benchmark name                              iterations   average time 
Phone.parse/1 with an Andorra phone number      100000   11.15 µs/op
Phone.parse/1 with a Zimbabwe phone number      100000   11.86 µs/op
Phone.parse/1 with an NANP phone number           5000   365.90 µs/op
```

To be made:
* Same type of changes to `Phone.NANP` module.
* Some kind of parallel validation for Area codes.

**Comment by fcevado** [comment]
NANP pattern matching:
```
## PhoneBench
benchmark name                              iterations   average time 
Phone.parse/1 with an Andorra phone number      100000   10.71 µs/op
Phone.parse/1 with a Zimbabwe phone number      100000   12.29 µs/op
Phone.parse/1 with an NANP phone number          10000   221.26 µs/op
```

**Comment by fcevado** [comment]
@aaronrenner I think right now we are a little better then before. I'll work on the area codes calling(i dont know if using Tasks or keep the pattern matching). tagging you so you can follow the results.

**Comment by aaronrenner** [comment]
Wow! That change to Phone.clear helped a lot! Could you cherry pick that change and the benchmark change and merge them in while we continue to figure out the best approach for NANP phone numbers?

File: lib/phone/nanp.ex [hunk 1]
```diff
@@ -1,32 +1,85 @@
...
-    Phone.NANP.TC,
-    Phone.NANP.TT,
-    Phone.NANP.US,
-    Phone.NANP.VC,
-    Phone.NANP.VG,
-    Phone.NANP.VI,
-    Phone.NANP.TollFree
-  ]
-  match :modules
+  def build("1684" <> _ = number), do: Phone.NANP.AS.build(number)
```
Comment (aaronrenner) [inline lib/phone/nanp.ex]: It's cool to see the performance improvement for these, but it's a bummer that these prefixes have to be duplicated everywhere. Could the `build/1` and `build!/1` functions be written using some sort of macro?

**Comment by aaronrenner** [comment]
I think you have a much faster computer than me. 😄  I just ran the benchmarks before and after the `Phone.clear` change, and didn't see much for performance difference.

Before:
```
phone HEAD % mix bench
Settings:
  duration:      1.0 s

## PhoneBench
[12:49:47] 1/3: Phone.parse/1 with a Zimbabwe phone number
[12:49:51] 2/3: Phone.parse/1 with an Andorra phone number
[12:49:54] 3/3: Phone.parse/1 with an NANP phone number

Finished in 8.25 seconds

## PhoneBench
benchmark name                              iterations   average time
Phone.parse/1 with an Andorra phone number        2000   750.36 µs/op
Phone.parse/1 with a Zimbabwe phone number        5000   758.01 µs/op
Phone.parse/1 with an NANP phone number           1000   1210.56 µs/op
```

After:
```
phone ar-phone-clear-performance % mix bench
Settings:
  duration:      1.0 s

## PhoneBench
[12:52:29] 1/3: Phone.parse/1 with a Zimbabwe phone number
[12:52:31] 2/3: Phone.parse/1 with an Andorra phone number
[12:52:33] 3/3: Phone.parse/1 with an NANP phone number

Finished in 5.91 seconds

## PhoneBench
benchmark name                              iterations   average time
Phone.parse/1 with a Zimbabwe phone number        2000   724.15 µs/op
Phone.parse/1 with an Andorra phone number        2000   731.53 µs/op
Phone.parse/1 with an NANP phone number           1000   1174.76 µs/op
```

I'll keep looking at this in my spare time, too.

File: lib/phone/nanp.ex (hunk 1 above)
Comment (fcevado) [inline lib/phone/nanp.ex]: that's what I'm thinking to work right now. I had some ideas and will try something this weekend. The way I'm thinking of doind this, pattern match will work for countries and area codes.

**Comment by fcevado** [comment]
> Wow! That change to Phone.clear helped a lot!

Actually the changes that improved the performance was made on the `Phone.Countries` module. :stuck_out_tongue: 
I didn't measure the impact of the changes on `Phone.clear`.

**Pull request closed without merging**